### PR TITLE
add spawn file argument to frida-inject

### PR DESCRIPTION
**Pull request by hunterli**
./frida-inject -f com.android.browser -s hello.js

### CODE REVIEW

File: inject/inject.vala [hunk 1]
```diff
@@ -11,8 +12,9 @@ namespace Frida.Inject {
 	private static bool output_version;
 
 	const OptionEntry[] options = {
-		{ "pid", 'p', 0, OptionArg.INT, ref target_pid, null, "PID" },
-		{ "name", 'n', 0, OptionArg.STRING, ref target_name, null, "PID" },
+		{ "pid", 'p', 0, OptionArg.INT, ref target_pid, "attach to PID", "PID" },
+		{ "name", 'n', 0, OptionArg.STRING, ref target_name, "attach to NAME", "NAME" },
+		{ "file", 'f', 0, OptionArg.STRING, ref spawn_file, "spawn FILE", "FILE" },
```
Comment (oleavr) [inline inject/inject.vala]: Let's add this one first in the list, for consistency with `frida-tools`.

File: inject/inject.vala [hunk 2]
```diff
@@ -45,7 +47,7 @@ namespace Frida.Inject {
 			return 1;
 		}
 
-		if (target_pid == -1 && target_name == null) {
+		if (target_pid == -1 && target_name == null && spawn_file == null) {
```
Comment (oleavr) [inline inject/inject.vala]: And move it first here as well.

File: inject/inject.vala [hunk 3]
```diff
@@ -135,6 +137,11 @@ namespace Frida.Inject {
 			construct;
 		}
 
+		public string? spawn_file {
+			get;
+			construct;
+		}
```
Comment (oleavr) [inline inject/inject.vala]: And here.

File: inject/inject.vala [hunk 4]
```diff
@@ -168,11 +175,12 @@ namespace Frida.Inject {
 		private int exit_code;
 		private MainLoop loop;
 
-		public Application (int target_pid, string? target_name, string? script_path, string? script_source,
+		public Application (int target_pid, string? target_name, string? spawn_file, string? script_path, string? script_source,
```
Comment (oleavr) [inline inject/inject.vala]: And here.

File: inject/inject.vala [hunk 5]
```diff
@@ -168,11 +175,12 @@ namespace Frida.Inject {
...
 				ScriptRuntime script_runtime, Json.Node parameters, bool enable_development) {
 			Object (
 				target_pid: target_pid,
 				target_name: target_name,
+				spawn_file: spawn_file,
```
Comment (oleavr) [inline inject/inject.vala]: And here.

File: inject/inject.vala [hunk 6]
```diff
@@ -202,7 +210,9 @@ namespace Frida.Inject {
 				var device = yield device_manager.get_device_by_type (DeviceType.LOCAL, 0, io_cancellable);
 
 				uint pid;
-				if (target_name != null) {
+				if (spawn_file != null) {
+					pid = yield device.spawn(spawn_file, null, io_cancellable);
```
Comment (oleavr) [inline inject/inject.vala]:
```suggestion
					pid = yield device.spawn (spawn_file, null, io_cancellable);
```

Comment (oleavr) [reply inject/inject.vala]: Should also `device.resume ()` here after loading the script.

File: inject/inject.vala (hunk 1 above)
Comment (hunterli) [reply inject/inject.vala]: OK

File: inject/inject.vala (hunk 2 above)
Comment (hunterli) [reply inject/inject.vala]: OK

File: inject/inject.vala (hunk 3 above)
Comment (hunterli) [reply inject/inject.vala]: OK

File: inject/inject.vala (hunk 4 above)
Comment (hunterli) [reply inject/inject.vala]: OK

File: inject/inject.vala (hunk 5 above)
Comment (hunterli) [reply inject/inject.vala]: OK

**Comment by oleavr** [comment]
Thanks!

**Pull request merged**